### DISCUSSION TITLE
Keep commit hashes when migrating to LFS?

### DISCUSSION THREAD

**Discussion by borekb** [git-lfs/git-lfs#4277] [Ideas] [answered]
(I know this is more about Git itself but the use case is centered around LFS and I'd like to know what you think.)

So the use case is that I want to migrate to Git LFS via `git lfs migrate import`. My options are:

1. Create a new commit via `--no-rewrite`.
    - (+) All the previous commit hashes are intact.
    - (-) My repo is still many GB in size (or hundreds of GB).
2. Rewrite all the history and force-push 
    - (+) The repo becomes much lighter 🎉
    - (-) All the PR discussions, CI builds, artifact versions (that are based on Git hashes), links from Slack, emails etc. become broken 😕

The no. 2 is so disruptive that we cannot use it in practice. But having a large repo is also a pain, so I'm fantasizing about a third option which would work like this:

- Git itself implements some mechanism that allows it to understand that the contents of an object (the blob) moved to LFS.
- The `migrate` command uses this mechanism.

_"But wait, Git objects are SHA-1 checksums of their contents, this can never work."_ I know but how unimaginable it really is? Git uses hashes for several good reasons, like being good identifiers in a distributed system, being able to verify objects' integrity (resiliency to network / disk failures), no one can maliciously mess with the history, etc. So would it be possible to keep these benefits?

I have something like this in mind (pseudocode):

```
$ git cat-file -p f030eaa00360064f3937db1be2d8d30ffcc310e
<<moved-to-lfs>>
<<signed-by: ...>>
version https://git-lfs.github.com/spec/v1
oid sha256:397d03745813ed042eadf668c65ba383201446db67924f056c435a25f1f3bd5e
size 184284
```

I don't know exactly but to me, it seems that Git _could_ learn to trust such objects. This itself wouldn't be enough, for example, for push/pull scenarios, there would need to be some way to indicate that the contents of an object ID changed and that the exchange should happen, I'm just trying to find out how "impossible" it is for Git to have objects in its database where their IDs are not necessarily SHA-1 checksums of their contents.

Because to me, from a user perspective, it would be _awesome_ to be able to make my repo much lighter while not breaking many other things that depend on commit IDs.

**Comment by bk2204** [comment, marked as answer]
I think there are two possibilities of approaches that could be taken here based on things that are already in Git.

The first is grafts.  Git has support for these already, and they replace one object with another one.  Git normally would still download the original object, but perhaps making it support grafts more natively would be helpful.

The other is partial clone.  This is sort of different from using Git LFS altogether, but it's possible to use them in conjunction.  Because you would avoid downloading the blobs in the first place and only download them on demand, your large blobs wouldn't be on your system.  Do note that there are cases where partial clone isn't very efficient, so it may not be the best option.

In general, Git upstream doesn't like to accept features that are Git LFS specific, because it tries not to prioritize any existing tool (and there are tools which compete with it, such as git-annex).  So if you're looking for Git to produce an LFS-specific solution, then that isn't likely to be accepted.

**Reply by borekb** [reply]
Wow, it's the first time I hear about grafts (probably replaced by [`replace`](https://git-scm.com/docs/git-replace) if I understand correctly). I have no experience with this yet, would it allow us to replace "commits referencing local blobs" with "commits referencing LFS blobs"? It sounds like it could but also a bit too good to be true 😄.

**Reply by bk2204** [reply]
It's possible to do, but I don't believe that `git replace` is enabled by default and I don't know that it prevents the download of the replaced objects.  It's just not an area that I'm very familiar with in Git, so you'd need to do more research to figure out if it's viable for your use case.

**Reply by borekb** [reply]
Yep, will look into it. Do you generally think that having a `migrate import` command that maintains commit hashes would be useful?

**Reply by bk2204** [reply]
Because I'm pretty sure that `git replace` isn't robust enough for this purpose, I'm going to recommend that we don't add such a mode at the moment.  If in the future it becomes robust enough and can be enabled by default for users with sane behavior, then we can accept a PR that adds such support.

**Reply by borekb** [reply]
Yeah, I'm just amazed that there's such a workflow in Git at all. I'm exploring it, it can indeed replace commits but I'm not sure if I can safely remove the old ones, for example, it seems like `git prune` doesn't remove them automatically and deleting `.git/objects` files manually feels wrong. But maybe I'm doing something wrong, I need to play more with it.

Thanks for the pointers though!
